// ng2-drag-and-check/ng2-drag-and-check.es5.js
import { Directive, ElementRef, EventEmitter, HostListener, Input, NgModule, Output, Renderer } from '@angular/core';
var Position = (function () {
    /**
     * @param {?} x
     * @param {?} y
     */
    function Position(x, y) {
        this.x = x;
        this.y = y;
    }
    return Position;
}());
var DragAndCheckDirective = (function () {
    /**
     * @param {?} el
     * @param {?} renderer
     */
    function DragAndCheckDirective(el, renderer) {
        this.el = el;
        this.renderer = renderer;
        this.allowDrag = true;
        this.moving = false;
        this.original = null;
        this.oldTrans = new Position(0, 0);
        this.tempTrans = new Position(0, 0);
        this.currentTrans = new Position(0, 0);
        this.oldZIndex = '';
        this.oldPosition = '';
        this.started = new EventEmitter();
        this.stopped = new EventEmitter();
        this.edge = new EventEmitter();
        this.allowedOffsets = new Offsets();
    }
    Object.defineProperty(DragAndCheckDirective.prototype, "ngDraggable", {
        /**
         * @param {?} setting
         * @return {?}
         */
        set: function (setting) {
            if (setting !== undefined && setting !== null && setting !== '') {
                this.allowDrag = !!setting;
                var /** @type {?} */ element = this.handle ? this.handle : this.el.nativeElement;
                if (this.allowDrag) {
                    this.renderer.setElementClass(element, 'ng-draggable', true);
                }
                else {
                    this.renderer.setElementClass(element, 'ng-draggable', false);
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DragAndCheckDirective.prototype.ngOnInit = function () {
        if (this.allowDrag) {
            var /** @type {?} */ element = this.handle ? this.handle : this.el.nativeElement;
            this.renderer.setElementClass(element, 'ng-draggable', true);
        }
    };
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    DragAndCheckDirective.prototype.getPosition = function (x, y) {
        return new Position(x, y);
    };
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    DragAndCheckDirective.prototype.moveTo = function (x, y) {
        if (this.original) {
            var /** @type {?} */ element = this.el.nativeElement;
            this.tempTrans.x = x - this.original.x;
            this.tempTrans.y = y - this.original.y;
            this.verifyMoveOnX();
            this.verifyMoveOnY();
            var /** @type {?} */ value = "translate(" + (this.tempTrans.x + this.oldTrans.x) + "px, " + (this.tempTrans.y + this.oldTrans.y) + "px)";
            this.renderer.setElementStyle(element, 'transform', value);
            this.renderer.setElementStyle(element, '-webkit-transform', value);
            this.renderer.setElementStyle(element, '-ms-transform', value);
            this.renderer.setElementStyle(element, '-moz-transform', value);
            this.renderer.setElementStyle(element, '-o-transform', value);
            this.edge.emit(this.boundsCheck());
            this.currentTrans = this.getPosition(this.tempTrans.x, this.tempTrans.y);
        }
    };
    /**
     * @return {?}
     */
    DragAndCheckDirective.prototype.verifyMoveOnX = function () {
        var /** @type {?} */ boundaries = this.el.nativeElement.getBoundingClientRect();
        var /** @type {?} */ horizontalMove = this.tempTrans.x - this.currentTrans.x;
        var /** @type {?} */ allowedLeftOffset = this.allowedOffsets.getLeft(this.el.nativeElement, this.handle);
        if (allowedLeftOffset != null &&
            horizontalMove < 0 &&
            boundaries.left + horizontalMove < allowedLeftOffset) {
            this.tempTrans.x = this.currentTrans.x - (boundaries.left - allowedLeftOffset);
        }
        var /** @type {?} */ allowedRightOffset = this.allowedOffsets.getRight(this.el.nativeElement, this.handle);
        var /** @type {?} */ windowWidth = document.documentElement.clientWidth;
        if (allowedRightOffset != null &&
            horizontalMove > 0 &&
            boundaries.right + horizontalMove > windowWidth - allowedRightOffset) {
            this.tempTrans.x = this.currentTrans.x + (windowWidth - allowedRightOffset - boundaries.right);
        }
    };
    /**
     * @return {?}
     */
    DragAndCheckDirective.prototype.verifyMoveOnY = function () {
        var /** @type {?} */ boundaries = this.el.nativeElement.getBoundingClientRect();
        var /** @type {?} */ windowHeight = document.documentElement.clientHeight;
        var /** @type {?} */ verticalMove = this.tempTrans.y - this.currentTrans.y;
        var /** @type {?} */ allowedTopOffset = this.allowedOffsets.getTop(this.el.nativeElement, this.handle);
        if (allowedTopOffset != null &&
            verticalMove < 0 &&
            boundaries.top + verticalMove < allowedTopOffset) {
            this.tempTrans.y = this.currentTrans.y - (boundaries.top - allowedTopOffset);
        }
        var /** @type {?} */ allowedBottomOffset = this.allowedOffsets.getBottom(this.el.nativeElement, this.handle);
        if (allowedBottomOffset != null &&
            verticalMove > 0 &&
            boundaries.bottom + verticalMove > windowHeight - allowedBottomOffset) {
            this.tempTrans.y = this.currentTrans.y + (windowHeight - allowedBottomOffset - boundaries.bottom);
        }
    };
    /**
     * @return {?}
     */
    DragAndCheckDirective.prototype.pickUp = function () {
        // get old z-index and position:
        this.oldZIndex = this.el.nativeElement.style.zIndex ? this.el.nativeElement.style.zIndex : '';
        this.oldPosition = this.el.nativeElement.style.position ? this.el.nativeElement.style.position : '';
        this.getPosition(0, 0);
        if (window) {
            this.oldZIndex = window.getComputedStyle(this.el.nativeElement, null).getPropertyValue('z-index');
            this.oldPosition = window.getComputedStyle(this.el.nativeElement, null).getPropertyValue('position');
        }
        // setup default position:
        var /** @type {?} */ position = 'relative';
        // check if old position is draggable:
        if (this.oldPosition && (this.oldPosition === 'absolute' ||
            this.oldPosition === 'fixed' ||
            this.oldPosition === 'relative')) {
            position = this.oldPosition;
        }
        this.renderer.setElementStyle(this.el.nativeElement, 'position', position);
        this.renderer.setElementStyle(this.el.nativeElement, 'z-index', '99999');
        if (!this.moving) {
            this.started.emit(this.el.nativeElement);
            this.moving = true;
        }
    };
    /**
     * @return {?}
     */
    DragAndCheckDirective.prototype.boundsCheck = function () {
        if (!this.bounds) {
            return {
                'top': false,
                'right': false,
                'bottom': false,
                'left': false
            };
        }
        var /** @type {?} */ boundary = this.bounds.getBoundingClientRect();
        var /** @type {?} */ elem = this.el.nativeElement.getBoundingClientRect();
        return {
            'top': boundary.top < elem.top,
            'right': boundary.right > elem.right,
            'bottom': boundary.bottom > elem.bottom,
            'left': boundary.left < elem.left
        };
    };
    /**
     * @return {?}
     */
    DragAndCheckDirective.prototype.putBack = function () {
        if (this.oldZIndex) {
            this.renderer.setElementStyle(this.el.nativeElement, 'z-index', this.oldZIndex);
        }
        else {
            this.el.nativeElement.style.removeProperty('z-index');
        }
        if (this.moving) {
            this.stopped.emit(this.el.nativeElement);
            this.edge.emit(this.boundsCheck());
            this.moving = false;
            this.oldTrans.x += this.tempTrans.x;
            this.oldTrans.y += this.tempTrans.y;
            this.tempTrans.x = this.tempTrans.y = 0;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DragAndCheckDirective.prototype.onMouseDown = function (event) {
        // 1. skip right click;
        // 2. if handle is set, the element can only be moved by handle
        if (event.button == 2 || (this.handle !== undefined && event.target !== this.handle)) {
            return;
        }
        this.original = this.getPosition(event.clientX, event.clientY);
        this.pickUp();
    };
    /**
     * @return {?}
     */
    DragAndCheckDirective.prototype.onMouseUp = function () {
        this.putBack();
    };
    /**
     * @return {?}
     */
    DragAndCheckDirective.prototype.onMouseLeave = function () {
        this.putBack();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DragAndCheckDirective.prototype.onMouseMove = function (event) {
        if (this.moving && this.allowDrag) {
            this.moveTo(event.clientX, event.clientY);
        }
    };
    /**
     * @return {?}
     */
    DragAndCheckDirective.prototype.onTouchEnd = function () {
        this.putBack();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DragAndCheckDirective.prototype.onTouchStart = function (event) {
        event.stopPropagation();
        event.preventDefault();
        if (this.handle !== undefined && event.target !== this.handle) {
            return;
        }
        this.original = this.getPosition(event.changedTouches[0].clientX, event.changedTouches[0].clientY);
        this.pickUp();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    DragAndCheckDirective.prototype.onTouchMove = function (event) {
        event.stopPropagation();
        event.preventDefault();
        if (this.moving && this.allowDrag) {
            this.moveTo(event.changedTouches[0].clientX, event.changedTouches[0].clientY);
        }
    };
    return DragAndCheckDirective;
}());
DragAndCheckDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngDraggable]'
            },] },
];
/**
 * @nocollapse
 */
DragAndCheckDirective.ctorParameters = function () { return [
    { type: ElementRef, },
    { type: Renderer, },
]; };
DragAndCheckDirective.propDecorators = {
    'started': [{ type: Output },],
    'stopped': [{ type: Output },],
    'edge': [{ type: Output },],
    'handle': [{ type: Input },],
    'bounds': [{ type: Input },],
    'allowedOffsets': [{ type: Input },],
    'ngDraggable': [{ type: Input },],
    'onMouseDown': [{ type: HostListener, args: ['mousedown', ['$event'],] },],
    'onMouseUp': [{ type: HostListener, args: ['document:mouseup',] },],
    'onMouseLeave': [{ type: HostListener, args: ['document:mouseleave',] },],
    'onMouseMove': [{ type: HostListener, args: ['document:mousemove', ['$event'],] },],
    'onTouchEnd': [{ type: HostListener, args: ['document:touchend',] },],
    'onTouchStart': [{ type: HostListener, args: ['touchstart', ['$event'],] },],
    'onTouchMove': [{ type: HostListener, args: ['document:touchmove', ['$event'],] },],
};
var Offsets = (function () {
    /**
     * @param {?=} top
     * @param {?=} right
     * @param {?=} bottom
     * @param {?=} left
     */
    function Offsets(top, right, bottom, left) {
        if (top === void 0) { top = Offsets.NONE; }
        if (right === void 0) { right = Offsets.NONE; }
        if (bottom === void 0) { bottom = Offsets.NONE; }
        if (left === void 0) { left = Offsets.NONE; }
        this.top = top;
        this.right = right;
        this.bottom = bottom;
        this.left = left;
    }
    /**
     * @param {?} el
     * @param {?=} handle
     * @return {?}
     */
    Offsets.prototype.getTop = function (el, handle) {
        return this.computeOffset(this.top, el, handle);
    };
    /**
     * @param {?} el
     * @param {?=} handle
     * @return {?}
     */
    Offsets.prototype.getRight = function (el, handle) {
        return this.computeOffset(this.right, el, handle);
    };
    /**
     * @param {?} el
     * @param {?=} handle
     * @return {?}
     */
    Offsets.prototype.getBottom = function (el, handle) {
        return this.computeOffset(this.bottom, el, handle);
    };
    /**
     * @param {?} el
     * @param {?=} handle
     * @return {?}
     */
    Offsets.prototype.getLeft = function (el, handle) {
        return this.computeOffset(this.left, el, handle);
    };
    /**
     * @param {?} value
     * @param {?} el
     * @param {?} handle
     * @return {?}
     */
    Offsets.prototype.computeOffset = function (value, el, handle) {
        var /** @type {?} */ result = null;
        if (!isNaN(+value)) {
            result = +value;
        }
        else {
            switch (value) {
                case Offsets.BORDER: {
                    result = 0;
                    break;
                }
                case Offsets.HALF_WIDTH: {
                    result = this.computeHALF(el);
                    break;
                }
                case Offsets.HALF_WIDTH: {
                    result = this.computeHALF(el);
                    break;
                }
                case Offsets.HANDLE_HEIGHT: {
                    result = this.computeHANDLE(el, handle);
                    break;
                }
                default:
            }
        }
        return result;
    };
    /**
     * @param {?} el
     * @return {?}
     */
    Offsets.prototype.computeHALF = function (el) {
        return -el.getBoundingClientRect().width / 2;
    };
    /**
     * @param {?} el
     * @param {?} handle
     * @return {?}
     */
    Offsets.prototype.computeHANDLE = function (el, handle) {
        if (!handle) {
            return 0;
        }
        var /** @type {?} */ contentHeight = el.getBoundingClientRect().height;
        return handle.getBoundingClientRect().height - contentHeight;
    };
    return Offsets;
}());
Offsets.NONE = 'none';
Offsets.BORDER = 'border';
Offsets.HALF_WIDTH = 'half';
Offsets.HANDLE_HEIGHT = 'handle';
var DragAndCheckModule = (function () {
    function DragAndCheckModule() {
    }
    return DragAndCheckModule;
}());
DragAndCheckModule.decorators = [
    { type: NgModule, args: [{
                declarations: [DragAndCheckDirective],
                exports: [DragAndCheckDirective]
            },] },
];
/**
 * @nocollapse
 */
DragAndCheckModule.ctorParameters = function () { return []; };
/**
 * Generated bundle index. Do not edit.
 */
export { DragAndCheckModule, DragAndCheckDirective, Offsets };
