// angular-calendar/esm5/date-adapters/date-adapter.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
var /**
 * @abstract
 */
DateAdapter = /** @class */ (function () {
    function DateAdapter() {
    }
    return DateAdapter;
}());
/**
 * @abstract
 */
export { DateAdapter };
if (false) {
    /**
     * @abstract
     * @param {?} date
     * @param {?} amount
     * @return {?}
     */
    DateAdapter.prototype.addWeeks = function (date, amount) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} amount
     * @return {?}
     */
    DateAdapter.prototype.addMonths = function (date, amount) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} amount
     * @return {?}
     */
    DateAdapter.prototype.subDays = function (date, amount) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} amount
     * @return {?}
     */
    DateAdapter.prototype.subWeeks = function (date, amount) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} amount
     * @return {?}
     */
    DateAdapter.prototype.subMonths = function (date, amount) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.getISOWeek = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} dayOfMonth
     * @return {?}
     */
    DateAdapter.prototype.setDate = function (date, dayOfMonth) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} month
     * @return {?}
     */
    DateAdapter.prototype.setMonth = function (date, month) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} year
     * @return {?}
     */
    DateAdapter.prototype.setYear = function (date, year) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.getDate = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.getMonth = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.getYear = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} amount
     * @return {?}
     */
    DateAdapter.prototype.addDays = function (date, amount) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} amount
     * @return {?}
     */
    DateAdapter.prototype.addHours = function (date, amount) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} amount
     * @return {?}
     */
    DateAdapter.prototype.addMinutes = function (date, amount) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} amount
     * @return {?}
     */
    DateAdapter.prototype.addSeconds = function (date, amount) { };
    /**
     * @abstract
     * @param {?} dateLeft
     * @param {?} dateRight
     * @return {?}
     */
    DateAdapter.prototype.differenceInDays = function (dateLeft, dateRight) { };
    /**
     * @abstract
     * @param {?} dateLeft
     * @param {?} dateRight
     * @return {?}
     */
    DateAdapter.prototype.differenceInMinutes = function (dateLeft, dateRight) { };
    /**
     * @abstract
     * @param {?} dateLeft
     * @param {?} dateRight
     * @return {?}
     */
    DateAdapter.prototype.differenceInSeconds = function (dateLeft, dateRight) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.endOfDay = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.endOfMonth = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?=} options
     * @return {?}
     */
    DateAdapter.prototype.endOfWeek = function (date, options) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.getDay = function (date) { };
    /**
     * @abstract
     * @param {?} dateLeft
     * @param {?} dateRight
     * @return {?}
     */
    DateAdapter.prototype.isSameDay = function (dateLeft, dateRight) { };
    /**
     * @abstract
     * @param {?} dateLeft
     * @param {?} dateRight
     * @return {?}
     */
    DateAdapter.prototype.isSameMonth = function (dateLeft, dateRight) { };
    /**
     * @abstract
     * @param {?} dateLeft
     * @param {?} dateRight
     * @return {?}
     */
    DateAdapter.prototype.isSameSecond = function (dateLeft, dateRight) { };
    /**
     * @abstract
     * @param {...?} dates
     * @return {?}
     */
    DateAdapter.prototype.max = function (dates) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} hours
     * @return {?}
     */
    DateAdapter.prototype.setHours = function (date, hours) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?} minutes
     * @return {?}
     */
    DateAdapter.prototype.setMinutes = function (date, minutes) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.startOfDay = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.startOfMinute = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.startOfMonth = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @param {?=} options
     * @return {?}
     */
    DateAdapter.prototype.startOfWeek = function (date, options) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.getHours = function (date) { };
    /**
     * @abstract
     * @param {?} date
     * @return {?}
     */
    DateAdapter.prototype.getMinutes = function (date) { };
}
