// angular-calendar/esm2015/modules/common/calendar-date-formatter.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * The parameter type passed to the date formatter methods.
 * @record
 */
export function DateFormatterParams() { }
if (false) {
    /**
     * The date to format.
     * @type {?}
     */
    DateFormatterParams.prototype.date;
    /**
     * The users preferred locale.
     * @type {?|undefined}
     */
    DateFormatterParams.prototype.locale;
    /**
     * The start day number of the week
     * @type {?|undefined}
     */
    DateFormatterParams.prototype.weekStartsOn;
    /**
     * An array of day indexes (0 = sunday, 1 = monday etc) that will be hidden on the view
     * @type {?|undefined}
     */
    DateFormatterParams.prototype.excludeDays;
    /**
     * The number of days in a week. Can be used to create a shorter or longer week view.
     * The first day of the week will always be the `viewDate`
     * @type {?|undefined}
     */
    DateFormatterParams.prototype.daysInWeek;
}
/**
 * If using a completely custom date formatter then it should implement this interface.
 * @record
 */
export function CalendarDateFormatterInterface() { }
if (false) {
    /**
     * The month view header week day labels
     * @param {?} __0
     * @return {?}
     */
    CalendarDateFormatterInterface.prototype.monthViewColumnHeader = function (__0) { };
    /**
     * The month view cell day number
     * @param {?} __0
     * @return {?}
     */
    CalendarDateFormatterInterface.prototype.monthViewDayNumber = function (__0) { };
    /**
     * The month view title
     * @param {?} __0
     * @return {?}
     */
    CalendarDateFormatterInterface.prototype.monthViewTitle = function (__0) { };
    /**
     * The week view header week day labels
     * @param {?} __0
     * @return {?}
     */
    CalendarDateFormatterInterface.prototype.weekViewColumnHeader = function (__0) { };
    /**
     * The week view sub header day and month labels
     * @param {?} __0
     * @return {?}
     */
    CalendarDateFormatterInterface.prototype.weekViewColumnSubHeader = function (__0) { };
    /**
     * The week view title
     * @param {?} __0
     * @return {?}
     */
    CalendarDateFormatterInterface.prototype.weekViewTitle = function (__0) { };
    /**
     * The time formatting down the left hand side of the day view
     * @param {?} __0
     * @return {?}
     */
    CalendarDateFormatterInterface.prototype.weekViewHour = function (__0) { };
    /**
     * The time formatting down the left hand side of the day view
     * @param {?} __0
     * @return {?}
     */
    CalendarDateFormatterInterface.prototype.dayViewHour = function (__0) { };
    /**
     * The day view title
     * @param {?} __0
     * @return {?}
     */
    CalendarDateFormatterInterface.prototype.dayViewTitle = function (__0) { };
}
