// angular-confirmation-popover/esm5/confirmation-popover-options.provider.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function ConfirmationPopoverOptionsInterface() { }
function ConfirmationPopoverOptionsInterface_tsickle_Closure_declarations() {
    /**
     * The popover title
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.popoverTitle;
    /**
     * The popover message
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.popoverMessage;
    /**
     * The popover confirmation button text
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.confirmText;
    /**
     * The popover cancel button text
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.cancelText;
    /**
     * The popover confirm button type e.g. `success`, `danger` etc
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.confirmButtonType;
    /**
     * The popover cancel button type  e.g. `success`, `danger` etc
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.cancelButtonType;
    /**
     * The popover placement. Can be `top`, `bottom`, `left`, `right`
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.placement;
    /**
     * Which button to cancel. Can be either `confirm` or `cancel`
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.focusButton;
    /**
     * Whether to hide the confirmation button
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.hideConfirmButton;
    /**
     * Whether to hide the cancel button
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.hideCancelButton;
    /**
     * A custom CSS class to be added to the popover
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.popoverClass;
    /**
     * Whether to append the popover to the document body
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.appendToBody;
    /**
     * Swap the order of the confirm and cancel buttons
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.reverseButtonOrder;
    /**
     * Whether or not the popover should stay open when clicking outside it
     * @type {?|undefined}
     */
    ConfirmationPopoverOptionsInterface.prototype.closeOnOutsideClick;
}
var ConfirmationPopoverOptions = /** @class */ (function () {
    function ConfirmationPopoverOptions() {
        this.confirmText = 'Confirm';
        this.cancelText = 'Cancel';
        this.confirmButtonType = 'success';
        this.cancelButtonType = 'default';
        this.placement = 'top';
        this.hideConfirmButton = false;
        this.hideCancelButton = false;
        this.popoverClass = '';
        this.appendToBody = false;
        this.reverseButtonOrder = false;
        this.closeOnOutsideClick = true;
    }
    return ConfirmationPopoverOptions;
}());
export { ConfirmationPopoverOptions };
function ConfirmationPopoverOptions_tsickle_Closure_declarations() {
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.popoverTitle;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.popoverMessage;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.confirmText;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.cancelText;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.confirmButtonType;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.cancelButtonType;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.placement;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.focusButton;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.hideConfirmButton;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.hideCancelButton;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.popoverClass;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.appendToBody;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.reverseButtonOrder;
    /** @type {?} */
    ConfirmationPopoverOptions.prototype.closeOnOutsideClick;
}
